File: src/main/scala/streaming/WordCountOnline.scala
package main.scala.streaming

import org.apache.spark.SparkConf
import org.apache.spark.streaming.StreamingContext
import org.apache.spark.streaming.Durations
import org.apache.spark.storage.StorageLevel

object WordCountOnline {
  def main(args: Array[String]): Unit = {
    
    val sparkConf = new SparkConf()
    sparkConf.setMaster("local[2]")
    sparkConf.setAppName("WordCountOnline")
    
    
    val ssc = new StreamingContext(sparkConf,Durations.seconds(5))
    
    val linesDStream = ssc.socketTextStream("hadoop1", 9999, StorageLevel.MEMORY_AND_DISK)
    
    val wordsDStream = linesDStream.flatMap { _.split(" ") }
    
    val pairDStream = wordsDStream.map { (_,1) }
    
    val resultDStream = pairDStream.reduceByKey(_+_)
    
    resultDStream.print()
    
    ssc.start()
    ssc.awaitTermination()
    ssc.stop()
    
    
  }
}

File: src/main/java/streaming/test/WordCountOnLine.java
package main.java.streaming.test;

import java.util.Arrays;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import org.apache.spark.streaming.Durations;
import org.apache.spark.streaming.api.java.JavaDStream;
import org.apache.spark.streaming.api.java.JavaPairDStream;
import org.apache.spark.streaming.api.java.JavaReceiverInputDStream;
import org.apache.spark.streaming.api.java.JavaStreamingContext;

import scala.Tuple2;

public class WordCountOnLine {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf()
				.setMaster("local[2]")
				.setAppName("WordCountOnLine");
		JavaStreamingContext streamingContext = new JavaStreamingContext(conf, Durations.seconds(5));
		
		JavaReceiverInputDStream<String> linesDStram = streamingContext.socketTextStream("hadoop1", 9999);
		
		JavaDStream<String> wordsDStream = linesDStram.flatMap(new FlatMapFunction<String, String>() {

			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public Iterable<String> call(String str) throws Exception {
				return Arrays.asList(str.split(" "));
			}
		});
		
		
		JavaPairDStream<String, Integer> pairDStram = wordsDStream.mapToPair(new PairFunction<String, String, Integer>() {

			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public Tuple2<String, Integer> call(String t) throws Exception {
				return new Tuple2<String, Integer>(t,1);
			}
		});
		
		JavaPairDStream<String, Integer> resultDStream = pairDStram.reduceByKey(new Function2<Integer, Integer, Integer>() {
			
			/**
			 * 
			 */
			private static final long serialVersionUID = 1L;

			@Override
			public Integer call(Integer v1, Integer v2) throws Exception {
				return v1 + v2;
			}
		});
		
		resultDStream.print();
		
		streamingContext.start();
		streamingContext.awaitTermination();
		
		streamingContext.close();
	}
}
